libvips/foreign/jp2ksave.c

```diff
@@ -0,0 +1,1181 @@
+			vips_foreign_save_jp2k_rgb_to_ycc( jp2k, &tile );
+
+		if( jp2k->downsample )
+			vips_foreign_save_jp2k_unpack_downsample( jp2k, &tile );
+		else
+			vips_foreign_save_jp2k_unpack( jp2k, &tile );
+
+		sizeof_tile = vips_foreign_save_jp2k_sizeof_tile( jp2k, &tile );
+		tile_index = tiles_across * tile.top / jp2k->tile_height +
+			x / jp2k->tile_width;
+		if( !opj_write_tile( jp2k->codec, tile_index, 
+			(VipsPel *) jp2k->tile_buffer, sizeof_tile, 
+			jp2k->stream ) )
+			return( -1 );
+	}
+
+	return( 0 );
+}
+
+static int
+vips_foreign_save_jp2k_write_block( VipsRegion *region, VipsRect *area, 
+	void *a )
+{
+	VipsForeignSaveJp2k *jp2k = (VipsForeignSaveJp2k *) a;
+	VipsForeignSave *save = (VipsForeignSave *) jp2k;
+
+#ifdef DEBUG_VERBOSE
+	printf( "vips_foreign_save_jp2k_write_block: y = %d, nlines = %d\n", 
+		area->top, area->height );
+#endif /*DEBUG_VERBOSE*/
+
+	for(;;) {
+		VipsRect hit;
+		int y;
+		VipsRect strip_position;
+
+		/* The intersection with the strip is the fresh pixels we
+		 * have. 
+		 */
+		vips_rect_intersectrect( area, &(jp2k->strip->valid), &hit );
+
+		/* Copy the new pixels into the strip.
+		 */
+		for( y = 0; y < hit.height; y++ ) {
+			VipsPel *p = VIPS_REGION_ADDR( region, 
+				0, hit.top + y );
+			VipsPel *q = VIPS_REGION_ADDR( jp2k->strip, 
+				0, hit.top + y );
+
+			memcpy( q, p, VIPS_IMAGE_SIZEOF_LINE( region->im ) );
+		}
+
+		/* Have we failed to reach the bottom of the strip? We must
+		 * have run out of fresh pixels, so we are done.
+		 */
+		if( VIPS_RECT_BOTTOM( &hit ) != 
+			VIPS_RECT_BOTTOM( &jp2k->strip->valid ) ) 
+			break;
+
+		/* We have reached the bottom of the strip. Write this line of
+		 * pixels and ove the strip down.
+		 */
+		if( vips_foreign_save_jp2k_write_tiles( jp2k ) )
+			return( -1 );
+
+		strip_position.left = 0;
+		strip_position.top = jp2k->strip->valid.top + jp2k->tile_height;
+		strip_position.width = save->ready->Xsize;
+		strip_position.height = jp2k->tile_height;
+		if( vips_region_buffer( jp2k->strip, &strip_position ) )
+			return( -1 );
+	}
+
+	return( 0 );
+}
+
+static int
+vips_foreign_save_jp2k_build( VipsObject *object )
+{
+	VipsObjectClass *class = VIPS_OBJECT_GET_CLASS( object );
+	VipsForeignSave *save = (VipsForeignSave *) object;
+	VipsForeignSaveJp2k *jp2k = (VipsForeignSaveJp2k *) object;
+
+	OPJ_COLOR_SPACE color_space;
+	int expected_bands;
+	int bits_per_pixel;
+	int i;
+	size_t sizeof_tile;
+	size_t sizeof_line;
+	VipsRect strip_position;
+
+	if( VIPS_OBJECT_CLASS( vips_foreign_save_jp2k_parent_class )->
+		build( object ) )
+		return( -1 );
+
+	opj_set_default_encoder_parameters( &jp2k->parameters );
+
+	/* Analyze our arguments.
+	 */
+
+	if( !vips_band_format_isint( save->ready->BandFmt ) ) {
+		vips_error( class->nickname,
+			"%s", _( "not an integer format" ) );
+		return( -1 );
+	}
+
+	switch( jp2k->subsample_mode ) {
+	case VIPS_FOREIGN_SUBSAMPLE_AUTO:
+		jp2k->downsample =
+			!jp2k->lossless &&
+			jp2k->Q < 90 &&
```

Comment:
Should this be `<= 90` to match the docs?

Reply:
I'm so confused, I though 90 and over was no chroma subsampling, 89 was chroma subsampled. Let me check.

Reply:
OK, made it >= 90 everywhere. The docs, comments and code were all inconsistent :(